anvil/layers.py

```diff
@@ -0,0 +1,1188 @@
+
+from anvil.core import NeuralNetwork
+
+import numpy as np
+
+
+class CouplingLayer(nn.Module):
+    """
+    Base class for coupling layers.
+
+    A generic coupling transformation takes the form
+
+        x_a = \phi_a
+        x_b = C( \phi_b ; {N(\phi_a)} )
+
+    where the D-dimensional \phi vector has been split into two D/2-dimensional vectors
+    \phi_a and \phi_b, and {N(\phi_a)} is a set of functions, possible neural networks,
+    which take \phi_a as parameters.
+
+    Parameters
+    ----------
+    size_half: int
+        Half of the configuration size, which is the size of the input vector
+        for the neural networks.
+    even_sites: bool
+        dictates which half of the data is transformed as a and b, since
+        successive affine transformations alternate which half of the data is
+        passed through neural networks.
+
+
+    Attributes
+    ----------
+    a_ind: slice (protected)
+        Slice object which can be used to access the passive partition.
+    b_ind: slice (protected)
+        Slice object which can be used to access the partition that gets transformed.
+    join_func: function (protected)
+        Function which returns the concatenation of the two partitions in the
+        appropriate order.
+    """
+
+    def __init__(self, size_half: int, even_sites: bool):
+        super().__init__()
+
+        if even_sites:
+            # a is first half of input vector
+            self._a_ind = slice(0, size_half)
+            self._b_ind = slice(size_half, 2 * size_half)
+            self._join_func = torch.cat
+        else:
+            # a is second half of input vector
+            self._a_ind = slice(size_half, 2 * size_half)
+            self._b_ind = slice(0, size_half)
+            self._join_func = lambda a, *args, **kwargs: torch.cat(
+                (a[1], a[0]), *args, **kwargs
+            )
+
+
+class AdditiveLayer(CouplingLayer):
+    def __init__(
+        self,
+        i: int,
+        size_half: int,
+        *,
+        hidden_shape: list,
+        activation: str,
+        symmetric: bool,
+        even_sites: bool,
+    ):
+        super().__init__(size_half, even_sites)
+        self.i = i
+
+        self.t_network = NeuralNetwork(
+            size_in=size_half,
+            size_out=size_half,
+            hidden_shape=hidden_shape,
+            activation=activation,
+            final_activation=None,
+            symmetric=symmetric,
+        )
+
+    def forward(self, x_input, log_density, *unused) -> torch.Tensor:
+        r"""Forward pass of affine transformation."""
+        x_a = x_input[:, self._a_ind]
+        x_b = x_input[:, self._b_ind]
+        x_a_stand = (x_a - x_a.mean()) / x_a.std()  # reduce numerical instability
+
+        t_out = self.t_network(x_a_stand)
+
+        phi_b = x_b - t_out
+
+        phi_out = self._join_func([x_a, phi_b], dim=1)
+
+        if phi_out.requires_grad is False:
+            np.savetxt(f"layer_{self.i}.txt", phi_out)
```

Comment:
So these get saved in the sampling loop?

Reply:
Yeah this was just a quick addition to let me plot histograms of the intermediate states. The more sensible thing would be to save the output from each pair of coupling layers inside the modified `Sequential` if the user sets some flag to True. This can definitely go.

Reply:
That's fine, was just checking